### DISCUSSION TITLE
Supermodel1 replacement board + MISE + MIRE diagnostics

### DISCUSSION THREAD

**Discussion by neilbradley2084** [misterblack1/trs80-diagnosticrom#17] [Ideas]
I'm creating a Model I replacement board (if you'd like details, ask), and I'd like to include this diagnostic ROM in the image itself. Our project is open source GPLv2 so it should be compatible with the licensing, here.

In addition to permission to use, I'd also like to extend the diagnostics to include MISE and MIRE diagnostics, as well as our board's additional features (512K flash, 512K RAM, full graphics, programmable characters, etc...). All can be auto-detected, though I suspect with the diagnostic extensions it would no longer fit in a 2K ROM.

I'd rather cooperate with this project and am looking for guidance on engagement on how or what you'd like me to do, or even if you're interested at all.

FYI, As with Adrian, I'm in the Portland area, too.

**Comment by ki3v** [comment]
Hi!  This sounds like an interesting project. You’re welcome to use this code and I’d be interested in helping adapt it for your project. Looks like all 3 of us are in the Portland area. Cool!

I’m currently out of town until the 14th, and I know Adrian is under the weather.  Let’s figure out the most effective means of communication. 

How far along is your project?

Cheers,
Dave

**Reply by neilbradley2084** [reply]
Excellent! Well, if it matters, I'm one of the owners of Ground Kontrol in downtown and live in Tigard, and am a huge retrocomputing nerd and a fan of Adrian's channel. ;-)

The CPLD for video has been completed - both VGA and composite out. CPLD For decode and everything else is being worked on now. We're likely to order dev boards by the end of the week. Next steps for me is putting an emulator together for it (with all our hardware extensions), so... maybe a couple of weeks until we're powering on? I definitely have work to do!

Spec-wise, it's essentially this:

512K SRAM (16K pages)
512K Flash (16K pages)
CPU Speed can be 1.77/3.58, etc.. up through 14Mhz.
Powered by USB connector or standard power
80x25 video, as well as 64x16, and stipple graphics mode (like the Grafyx), memory mapped - VGA and composite
I/O mapped peripherals (like keyboard and FDC) so CP/M can run

We considered doing an expansion interface, but MISE kinda has that all taken care of. I have both a MIRE and a MISE, too.

**Comment by neilbradley2084** [comment]
Just got done with the proto board. This doesn't include the video circuitry (that's next) and it's not TRS-80 form factor - yet. We'll do that in the final design.

![image](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/eb51f24a-979b-44cb-a65e-44ca561c22db)

**Reply by Howlinmoon** [reply]
(if you'd like details, ask) - so I'm asking :-)

Will this have the standard 40 pin expansion bus, or is everything totally self contained?

I'd like to watch this develop - where is the project page hosted at?

Will it have cassette input/output?

**Reply by neilbradley2084** [reply]
It's a full size replacement board for the model 1, including the 40 pin expansion bus. It will have a cassette input/output. There isn't a web page for the project yet.

**Comment by misterblack1** [comment]
> I'm creating a Model I replacement board (if you'd like details, ask), and I'd like to include this diagnostic ROM in the image itself. Our project is open source GPLv2 so it should be compatible with the licensing, here.
> 
> In addition to permission to use, I'd also like to extend the diagnostics to include MISE and MIRE diagnostics, as well as our board's additional features (512K flash, 512K RAM, full graphics, programmable characters, etc...). All can be auto-detected, though I suspect with the diagnostic extensions it would no longer fit in a 2K ROM.

Yeah sounds good just email me and I can add David. We're not precious about the code -- so we could merge the changes back into the main brance. Sounds like an additional version would be needed to support these extra features. You thinking of just including it right into the system ROM so a user could chose to run diagnostics as needed?

**Reply by neilbradley2084** [reply]
Can do! Yes, the idea was to have the system be able to boot any of the level II ROM versions, level I, diagnostics, or our yet-to-be-decided custom ROM for the extended features. I'll do my best to analyze how the existing code works and do my best to cooperate with the existing architecture. I realize that the MIRE and MISE diagnostics are definitely extras, so I'll keep them segmented as much as possible, and that'll give us the flexibility we need to conditionally include them.

**Comment by ki3v** [comment]
I’d be happy to go over how the existing code works. It’s definitely unconventional, but that shouldn’t keep you from using conventional z80 code elsewhere. It ended up that way as a way to write non-trivial code in an environment where we did not yet trust any RAM or VRAM. 

I’m not a Z80 expert. This was my first Z80 project. But I’d be interested in helping out if possible. 

-
David Giller

**Reply by neilbradley2084** [reply]
Sounds great! I actually read your theory of operation and it does make a lot of sense. I've used several tricks you've used when doing BIOS development, back in the day when you had to run for a while before RAM got initialized and you couldn't use on-chip  cache as temp RAM. It makes 

Ultimately we'll want to detect the MISE/MIRE/SuperModel1 automatically and there are programmatic ways to detect their presence. If you have specific guidance on where I should look to adding said tests that'd be appreciated, or basically any hints on where stuff is or where it should be added to lessen the amount of work to include in the main tree. Ultimately I think MISE/MIRE should just have their own modules for testing what's on each device, but for the SuperModel1, it might make sense for that to be part of the mainline test.

**Comment by neilbradley2084** [comment]
For those interested, here's the updated/routed board:

![image](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/c26b34d0-ec57-4e4a-a9ee-0df2f14e0000)

**Reply by Howlinmoon** [reply]
Very nice looking!  Is there a specific list of features it will have?
I assume this: "programmable characters" means it will support lowercase with true descenders?
Does it have the option for 48K of ram in the keyboard - so you can yank the slower RAM in the old expansion interfaces?

**Reply by neilbradley2084** [reply]
So far I've just been copying this list around:

512K SRAM (16K pages)
512K Flash (16K pages)
CPU Speed can be 1.77/3.58, etc.. up through 14Mhz.
Powered by USB connector or standard power
80x25 video, as well as 64x16, and stipple graphics mode (like the Grafyx), memory mapped - VGA and composite
I/O mapped peripherals (like keyboard and FDC) so CP/M can run

Yes, all characters are programmable, so lower case and true descenders, including a memory mapped graphics (unlike the Grafyx) mode and 80x25. It has 512K of SRAM (no more dynamic!) in the keyboard so no need for memory in the EI (also, it being slow RAM wouldn't really matter since the TRS-80 didn't push it anywhere near that hard). The idea with the 512K is so you can page in 16K chunks of RAM (or flash). The idea also is to have multiple images in the flash so you can boot model 1 level 1, model 1 level II, built-in diagnostics, or whatever else we figure we'll put in it.

**Comment by ki3v** [comment]
How do you plan to get ROM images loaded?  What’s the plan for page mapping of ROM/RAM (in terms of how software/firmware will swap pages and what combinations of ROM/RAM paging will be possible), and for selecting the ROM to boot from?

If the memory map will need to be Model I compatible, and for booting an actual Model I ROM, I assume you’ll need to be able to map in a 2K page of ROM. In that case, the diagnostics should probably be a separate optional ROM or part of a larger mappable enhanced-mode ROM that gets mapped out when running Model I/III software. 

Any plans to support Model 4 software?  If the peripherals can alternately be memory (for Model I/III software) or I/O mapped, that might be not be far off from Model 4 compatibility?

**Reply by neilbradley2084** [reply]
Our existing plan is to have 4 bank switch registers, each for 16K of available address space. Something like this:

I/O ports 0x70-0x73: 0x70=Page 0 (0-16K), 0x71=Page 1 (16K-32K), etc...

Bits 0-5 - 16K Page within SRAM or flash (this is A
Bit 6 - 0=Read/writable, 1=Read only
Bit 7 - 0=SRAM, 1=Flash

Mapping peripherals like video, FDD, keyboard is separately controllable so you can map it in or out of the lowest region. Ideally we want to have an alternate method of access to FDD, keyboard (as I/O driven instead of memory mapped) so things like CP/M can run unrestricted.

Upon initial boot, the lowest part of flash will be mapped at 0. I had in mind to include all flash images into the flash, and reserve a 16K page for a boot loader/selector. Some sort of a startup sequence, like pressing and holding 1-6 on the keyboard while turning it on or resetting it to select which image you want to boot, or a boot loader menu that allows you to change it, but boots by default if no one touches it for a few seconds. Once the correct image is selected, code would just map it down to 0, disable interrupts, and jump to 0, so the hardware itself wouldn't be choosing the image to boot. The idea would be for it to remember which image to boot from in a nonvolatile fashion. Before any of the other ROMs operate unrestricted, the code will need to load up the TRS-80 character set into the video controller, too, so there will be some bootstrapping code required regardless.

There is a read protect there, too and there's a very specific reason why you can lock RAM regions. The ROM basic does do some destructive operations to itself - shifts on memory locations in ROM. When in ROM, the write gets completely ignored, but if it were RAM, it would be modified. So now anyone can map RAM into the lowest region, copy the ROM image down to RAM, then set it read only and voila - instant modified ROM.

I had my eye on the model 4 software. What I've come away with is we'd be required to use the CRTC to get 80 column mode (which we don't want to do, given our video subsystem is considerably better), the HDD controller is different, and the bank switching scheme is just... well, weird. The short of it is I didn't see a nice/easy way to deal with all of the model 4 specific oddities, but if you have a different opinion or specific guidance or ideas on how to make everything more compatible, I'm all ears!

**Reply by ki3v** [reply]
Haha the M4 banking is a little weird. I think I remember seeing it write to ROM space in the parts of the M4 ROM I disassembled… I think it does that when loading the Model III compatibility mode “ROM” from disk into the RAM under the ROM when it’s mapped.  Come to think about it, given that the M4 can emulate a M3, it might not be too hard to at least emulate a floppy-based M4. As far as the CRTC is concerned I would think you would only need to emulate the control registers, and you might need to do that anyway. But that’s just idle speculation… you’ve done a lot of work and I’m sure you’ve looked at it. 

I’m really happy the project will be open source. We will all be able to learn from your work.

**Reply by neilbradley2084** [reply]
Hey, don't give me too much credit! I'm certainly just getting started and am all ears if someone can come up with a better way, or a "more compatible" way. 

Additionally, the CPLD has a lot of marcocells left over, so a lot is still possible even if we lock everything down. The control registers in the CRTC aren't necessary for the 80x25 and graphics modes we're doing, but it's possible that all the code does currently is write to those registers, which we can handily ignore. 

But we're learning on this, too. I just started to piece together an emulator. Jeff managed to get the hardware CPLD generating characters last night in VGA mode.

![image](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/e4489444-1c0c-4210-a16e-3c7063e935db)
![image](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/31063d20-c6e5-4346-b13d-cb34d4257d79)

**Comment by neilbradley2084** [comment]
For those interested, there's a Discord for the Supermodel-1:

https://discord.com/channels/458077089493680160/1140065692960432128

**Reply by KelleyCook** [reply]
I’d love to join the discord, unfortunately the invite links will expire after one day.

**Comment by neilbradley2084** [comment]
Quick update... I did manage to get the emulator going, initial boards have been ordered as have parts. Exciting stuff!

![image](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/e492d912-e932-4f45-97e1-ae3bffcdd1dd)

**Reply by neilbradley2084** [reply]
And to get things on topic, the diagnostic ROM now runs unmodified! I had to implement several undocumented Z80 instructions...

![image](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/640012ee-d5b5-45f4-9c34-4c2cf4fef2a8)

**Reply by neilbradley2084** [reply]
Fixed the video:

![image](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/f325f7fa-13d8-4545-ab5a-81866ad0dac0)

**Comment by neilbradley2084** [comment]
Got the boot loader all done as well as the emulator to go along with it.
![TRS80](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/ab071301-f830-4feb-9fbd-2d24b5679c80)

**Comment by neilbradley2084** [comment]
A bit of an update from Jeff (HW designer) and me....

A small update on our supermodel1 project.  The first set of PCBs for the prototype are here and working great, and the video controller is also working with both NTSC and VGA, and no flicker or noise of any kind during CPU access even with the CPU at 16MHz.  CPU speed is software controllable from 1Mhz to 16Mhz.
Neil just came over today to grab a built up PCB set to work with, so we will have more cool ROM stuff soon. 
Both the NTSC and the VGA output look very clean, and the VGA output is in a 640x480 video mode with 512x384 actual pixels.
Next up is to test it with a real EI/Floppy Drive.
Has onboard RTC/NVRAM, so you can see boot parameter and it will remember them (boot up speed, date/time, etc)
Everything is open source/open hardware.  You could build your own with all through hole components you can order new from Digikey/Mouser today. 
Short video about it here:   [https://www.youtube.com/watch?v=S6-mbOI-b-U](https://www.youtube.com/watch?v=S6-mbOI-b-U&fbclid=IwAR3Dy0Es92hkbo41JKFhglpp6agJchJ0iKs6m62EYZBWtIT8MlMVkpoGlcs)

![t4](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/6dd6ce18-2b6c-4c24-8d88-92dcb6214c9d)
![t3](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/d477e79c-175b-4270-83c5-df727843a5b6)
![t2](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/7b9b923f-046b-44ba-8765-ceafc93f7532)
![t1](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/509caaf4-97fb-4834-bf55-94a1e4ed7fd2)
![t5](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/b6dca753-afc6-46f3-bb8f-f573260a48e7)

**Comment by neilbradley2084** [comment]
A bit of an update for everyone. We're going to include a microSD card so we'll emulate full floppy and hard disks, including the ability to format disks and (hopefully) have an operating system image we can image a disk.

https://www.youtube.com/watch?v=IqDobC2-UXg

**Comment by neilbradley2084** [comment]
And yet more of an update - we are finally doing board layout for the final form factor:

![image](https://github.com/misterblack1/trs80-diagnosticrom/assets/141876222/785af05e-d4f6-4aa0-b61a-f3565bc98091)
